src/systemd-sonic-generator/systemd-sonic-generator.c

```diff
@@ -128,6 +218,45 @@ static bool is_multi_instance_service(char *service_name){
 
 }
 
+
+/**
+ * Checks if a service is a multi-instance service for DPU.
+ *
+ * @param service_name The name of the service to check.
+ * @return true if the service is a multi-instance service for DPU, false otherwise.
+ */
+static bool is_multi_instance_service_for_dpu(const char *service_name) {
+    if (!smart_switch_npu) {
+        return false;
+    }
+
+    const static char* multi_instance_services_for_dpu[] = {"database"};
+    char *saveptr;
+    char *tmp_service_name = strdup(service_name);
```

Comment:
Maybe check if `tmp_service_name` is not NULL?

Reply:
done